src/OptiHPLCHandler/empower_instrument_method.py

```diff
@@ -144,6 +157,159 @@ class SampleManagerMethod(ColumnHandlerMethod):
+    """
+
+    valve_tag_prefix: str
+    valve_tag_suffix: str
+    solvent_lines: List[str]
+
+    def __init__(self, method_definition: Mapping[str, str]):
+        super().__init__(method_definition)
+        _gradient_xml = self.find_value(self.original_method["xml"], "GradientTable")
+        self.gradient_data = self.interpret_gradient_table(_gradient_xml)
+
+    @property
+    def current_method(self) -> DataModel:
+        """The current method definition, including the changes that have been made."""
+        method = super().current_method
+        original_gradient_xml = (
+            "<GradientTable>"
+            + self.find_value(self.original_method["xml"], "GradientTable")
+            + "</GradientTable>"
+        )
+        return self.alter_method(method, [(original_gradient_xml, self.gradient_xml)])
+
+    @property
+    def valve_position(self) -> List[str]:
+        """The current valve position for each solvent line."""
+        valve_position_tags = [
+            self.valve_tag_prefix + solvent + self.valve_tag_suffix
+            for solvent in self.solvent_lines
+        ]
+        valve_positions = [
+            solvent + self[tag]
+            for solvent, tag in zip(self.solvent_lines, valve_position_tags)
+        ]
+        # Consider removing the ones that have position 0,
+        # to make QSM methods easier to read.
+        return valve_positions
+
+    def __str__(self):
+        return f"{type(self).__name__} with valve positions {self.valve_position}"
+
+    @valve_position.setter
+    def valve_position(self, value: Union[str, List[str]]) -> None:
+        if isinstance(value, str):
+            value = [value]
+        for position in value:
+            if position[0] not in self.solvent_lines:
+                raise ValueError(
+                    f"Invalid valve position {position}, "
+                    f"must start with one of {self.solvent_lines}"
+                )
+            self[
+                self.valve_tag_prefix + position[0] + self.valve_tag_suffix
+            ] = position[1:]
+
+    @property
+    def gradient_table(self) -> List[Dict[str, str]]:
+        """
+        The gradient table for the method. It is a list of dicts, one for each row.
+
+        The dicts have the following keys:
+        - Time: The time in minutes.
+        - Flow: The flow in mL/min.
+        - CompositionX: The composition of solvent line X (A, B, C, D) in %.
+        - Curve: The curve type (1-11, 6 is linear and default).
+        """
+        gradient_table = []
+        for row in self.gradient_data:
+            row_dict = {"Time": row.time, "Flow": row.flow}
+            for solvent, composition in zip(self.solvent_lines, row.composition):
+                row_dict[f"Composition{solvent}"] = composition
+            row_dict["Curve"] = row.curve.value
+            gradient_table.append(row_dict)
+        return gradient_table
+
+    @gradient_table.setter
```

Comment:
Might get you to explain these next time we meet

Reply:
Getters and setters?  Sure thing, we will talk about that tomorrow.